### PR TITLE
[Traffic Capture] Add multipart support for offloading messages

### PR DESCRIPTION
**Pull request by lewijacn**
### Description
Some incoming packets will be larger than what our offloading protocols (e.g. Kafka) can handle.  This change chops those messages into pieces (while retaining proper protobuf format) and sends them to the offloader. The offloader will provide the `CodedOutputStream` and thus can set the max desired size of the message which will be determined by the size of the `CodedOutputStream` ~~and can also provide the minimum chunk size which will limit the maximum number of data bytes that will be chunked~~. This functionality required minimal change for offloaders (Kafka, File, InMemory) and was mainly rooted in the Serializer.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1061

### Testing
Manual testing with local kafka cluster and file directory and unit tests

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 1]
```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
...
+        // from a much larger message
+        int totalMessageBytesRequired = CodedOutputStreamSizeUtil.totalBytesNeededForMessage(timestamp,
+            segmentFieldNumber, segmentDataFieldNumber, segmentCountFieldNumber, 2, byteBuffer, numFlushesSoFar + 1);
+        int totalDataSurroundingBytesRequired = totalMessageBytesRequired - byteBuffer.capacity();
+
+        if (totalDataSurroundingBytesRequired + minDataChunkBytes >= getOrCreateCodedOutputStream().spaceLeft()) {
+            flushCommitAndResetStream(false);
+        }
+
+        if (byteBuffer.limit() == 0 || totalMessageBytesRequired <= getOrCreateCodedOutputStream().spaceLeft()) {
```
Comment (gregschohn) [inline TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:229]: why if byteBuffer.limit() == 0?  You should have a unit test to pick that one up.  Maybe I'm misremembering, but couldn't that cause overflow of your output?  If you have no bytes left & a 0 byte buffer, there's still overhead, right?

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 2]
```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
...
+            segmentFieldNumber = TrafficObservation.WRITESEGMENT_FIELD_NUMBER;
+            segmentCountFieldNumber = WriteSegment.COUNT_FIELD_NUMBER;
+            segmentDataFieldNumber = WriteSegment.DATA_FIELD_NUMBER;
+        }
+
+        // The required bytes here are not optimizing for space and instead are calculated on the worst case estimate of
+        // the potentially required bytes for simplicity. This could leave ~5 bytes of unused space in the CodedOutputStream
+        // when considering the case of a message that does not need segments or the case of a smaller segment created
+        // from a much larger message
+        int totalMessageBytesRequired = CodedOutputStreamSizeUtil.totalBytesNeededForMessage(timestamp,
```
Comment (gregschohn) [inline TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:221]: Since you modify this value in the while loop below, you should indicate that w/ an adjective like 'left'.  I'd also include something else like 'IncludingOverhead'

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 3]
```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
...
+
+        int dataCount = 0;
+        while(byteBuffer.position() < byteBuffer.limit()) {
+            int availableCOSSpace = getOrCreateCodedOutputStream().spaceLeft();
+            int chunkBytes = totalMessageBytesRequired > availableCOSSpace ? availableCOSSpace - totalDataSurroundingBytesRequired : byteBuffer.limit() - byteBuffer.position();
+            ByteBuffer bb = byteBuffer.slice(byteBuffer.position(), chunkBytes);
+            byteBuffer.position(byteBuffer.position() + chunkBytes);
+            addSubstreamMessage(segmentFieldNumber, segmentDataFieldNumber, segmentCountFieldNumber, ++dataCount, timestamp, bb);
+            // 1 to N-1 chunked messages
+            if (byteBuffer.position() < byteBuffer.limit()) {
```
Comment (gregschohn) [inline TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:242]: You'd also want to flush in situations where the next observation in was too big to fit in what's left.  If your max size is 1MB though, and the overhead is 20 bytes, that would be < 1 in 50K times - so it probably doesn't matter.

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 4]
```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
...
+
+        if (byteBuffer.limit() == 0 || totalMessageBytesRequired <= getOrCreateCodedOutputStream().spaceLeft()) {
+            addSubstreamMessage(observationFieldNumber, dataFieldNumber, timestamp, byteBuffer);
+            byteBuffer.position(byteBuffer.limit());
+        }
+
+        int dataCount = 0;
+        while(byteBuffer.position() < byteBuffer.limit()) {
+            int availableCOSSpace = getOrCreateCodedOutputStream().spaceLeft();
+            int chunkBytes = totalMessageBytesRequired > availableCOSSpace ? availableCOSSpace - totalDataSurroundingBytesRequired : byteBuffer.limit() - byteBuffer.position();
```
Comment (gregschohn) [inline TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:237]: Upon working through the arithmetic a couple times, I think that this might be right.  I'd like to see some more exhaustive unit tests that work through a number of edge conditions.  Starting w/ a smaller & buffer size and incrementing it by 1 for a sizeable range might do just that.

**Review by gregschohn — COMMENTED** [review]
Most of the questions are for discussion.  I don't understand the minDataSize value, nor do I think that the file or memory streams should be capped at 1MB.
One big blocker now though is on unit tests.  There's a lot of edge conditions in the arithmetic.  Dealing with byte buffers that may be pooled, especially when tossed in with some async activities, can create a number of issues as I've started to outline.  Tighten up the unit tests, checking the coverage on the new code would be helpful too to see if you've hit all of the edges/branches (both today & for any new ones).  Narrowly focused exhaustive testing seems apropos.

**Review by lewijacn — COMMENTED** [review]
Addressing some comments still some more I need to follow up on

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 5]
```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
...
+            // 1 to N-1 chunked messages
+            if (byteBuffer.position() < byteBuffer.limit()) {
+                flushCommitAndResetStream(false);
+                totalMessageBytesRequired = totalMessageBytesRequired - chunkBytes;
+            }
+        }
+
+    }
+
+    private void addSubstreamMessage(int observationFieldNumber, int dataFieldNumber, int dataCountFieldNumber, int dataCount,
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:251]: From looking at this there doesn't seem to be a simple division (which avoids a bunch of duplication) that makes sense as the order of writing kind of weaves them together. I think ultimately we could break out some functions and make this more logical but probably only makes sense if you have strong preference here

**Review by lewijacn — COMMENTED** [review]
Following up on rest of the comments and adding improvements

File: TrafficCapture/captureKafkaOffloader/src/main/java/org/opensearch/migrations/trafficcapture/kafkaoffloader/KafkaCaptureFactory.java [hunk 6]
```diff
@@ -70,7 +60,9 @@ public IChannelConnectionCaptureSerializer createOffloader(String connectionId)
                     CompletableFuture cf = new CompletableFuture<>();
                     // Async request to Kafka cluster
                     producer.send(record, handleProducerRecordSent(cf, recordId));
-                    return cf;
+                    // A more desirable way to cut off our tree of cf aggregation should be investigated
+                    aggregateCf.set(aggregateCf.get().isDone() ? cf : CompletableFuture.allOf(aggregateCf.get(), cf));
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/captureKafkaOffloader/src/main/java/org/opensearch/migrations/trafficcapture/kafkaoffloader/KafkaCaptureFactory.java]: The order of messages processed by Kafka.  Each one will fire independent of previous ones that were fired.  We'll only have the last close() call be aware of when all of the pieces have been committed (in some order).

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java (hunk 3 above)
Comment (gregschohn) [reply TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:274]: Don't worry about it.  I was thinking about speculating ahead - like if any of the next messages would be too big too, then flush now before the next round.  However, I don't think that it's worth it to make that optimization at this point.

File: TrafficCapture/captureKafkaOffloader/src/main/java/org/opensearch/migrations/trafficcapture/kafkaoffloader/KafkaCaptureFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lewijacn) [reply TrafficCapture/captureKafkaOffloader/src/main/java/org/opensearch/migrations/trafficcapture/kafkaoffloader/KafkaCaptureFactory.java]: Gotcha yes the Kafka messages are firing independent of each other but our CompletableFuture is synchronously expanding so that the final close() will be a future comprised of the previous ones

**Comment by lewijacn** [comment]
Updated PR after some discussion with @gregschohn on min data chunks and setting buffer size. Adding a follow up task here to add some additional edge testing that is needed here: https://opensearch.atlassian.net/browse/MIGRATIONS-1097

**Review by gregschohn — APPROVED** [review]
